src/distribution_univariate.F90

```diff
@@ -0,0 +1,343 @@
+    this%p(:) = p(:)
+
+    ! Normalize density function
+    this%p(:) = this%p(:)/sum(this%p)
+  end subroutine
+
+  function uniform_sample(this) result(x)
+    class(Uniform), intent(in) :: this
+    real(8) :: x
+
+    x = this%a + prn()*(this%b - this%a)
+  end function uniform_sample
+
+  function maxwell_sample(this) result(x)
+    class(Maxwell), intent(in) :: this
+    real(8) :: x
+
+    x = maxwell_spectrum(this%theta)
+  end function maxwell_sample
+
+  function watt_sample(this) result(x)
+    class(Watt), intent(in) :: this
+    real(8) :: x
+
+    x = watt_spectrum(this%a, this%b)
+  end function watt_sample
+
+  function tabular_sample(this) result(x)
+    class(Tabular), intent(in) :: this
+    real(8) :: x
+
+    integer :: i
+    real(8) :: c         ! sampled cumulative frequency
+    real(8) :: m         ! slope of PDF
+    real(8) :: x_i, x_i1 ! i-th and (i+1)th x values
+    real(8) :: c_i, c_i1 ! i-th and (i+1)th cumulative distribution values
+    real(8) :: p_i, p_i1 ! i-th and (i+1)th probability density values
+
+    ! Sample value of CDF
+    c = prn()
+
+    ! Find first CDF bin which is above the sampled value
+    c_i = this%c(1)
+    do i = 1, size(this%c) - 1
+      c_i1 = this%c(i + 1)
+      if (c <= c_i1) exit
+      c_i = c_i1
+    end do
+
+    ! Determine bounding PDF values
+    x_i = this%x(i)
+    p_i = this%p(i)
+
+    if (this%interpolation == HISTOGRAM) then
+      ! Histogram interpolation
+      if (p_i > ZERO) then
+        x = x_i + (c - c_i)/p_i
+      else
+        x = x_i
+      end if
+    else
+      ! Linear-linear interpolation
+      x_i1 = this%x(i + 1)
+      p_i1 = this%p(i + 1)
+
+      m = (p_i1 - p_i)/(x_i1 - x_i)
+      if (m == ZERO) then
+        x = x_i + (c - c_i)/p_i
+      else
+        x = x_i + (sqrt(max(ZERO, p_i*p_i + 2*m*(c - c_i))) - p_i)/m
+      end if
+    end if
+  end function tabular_sample
+
+  subroutine tabular_initialize(this, x, p, interp)
+    class(Tabular), intent(inout) :: this
+    real(8), intent(in) :: x(:)
+    real(8), intent(in) :: p(:)
+    integer, intent(in) :: interp
+
+    integer :: n
+
+    ! Check interpolation parameter
+
+    if (interp /= HISTOGRAM .and. interp /= LINEAR_LINEAR) then
+      call fatal_error('Only histogram and linear-linear interpolation for tabular &
+           &distribution is supported.')
+    end if
+
+    ! Check length of x, p arrays
+    if (size(x) /= size(p)) then
+      call fatal_error('Tabulated probabilities not of same length as &
+           &independent variable.')
+    end if
+
+    ! Copy probability density function and interpolation parameter
+    n = size(x)
+    allocate(this%x(n), this%p(n), this%c(n))
+    this%interpolation = interp
+    this%x(:) = x(:)
+    this%p(:) = p(:)
+
+    ! Calculate cumulative distribution function
+    this%c(1) = ZERO
+    do i = 2, n
+      if (this%interpolation == HISTOGRAM) then
+        this%c(i) = this%c(i-1) + this%p(i-1)*(this%x(i) - this%x(i-1))
```

Comment:
Yes, the probability of landing in a certain bin is `p_i * width_i`. The reason is that the `p_i`s then form a true probability density function in the sense that if you integrate it over the full domain you'll get one. Another way to think is that if you wanted to plot the probability density function you would do a step plot of `p_i` versus `x_i`.